packages/mui-base/src/utils/useScrollLock.ts

```diff
@@ -1,75 +1,104 @@
  *
  * @param enabled - Whether to enable the scroll lock.
  */
 export function useScrollLock(enabled: boolean = true) {
-  // Based on Floating UI's FloatingOverlay
-
   const lockId = useId();
+
   useEnhancedEffect(() => {
-    if (!enabled) {
+    if (!enabled || !lockId) {
       return undefined;
     }
 
-    activeLocks.add(lockId!);
-
-    const rootStyle = document.documentElement.style;
-    // RTL <body> scrollbar
-    const scrollbarX =
-      Math.round(document.documentElement.getBoundingClientRect().left) +
-      document.documentElement.scrollLeft;
-    const paddingProp = scrollbarX ? 'paddingLeft' : 'paddingRight';
+    const html = document.documentElement;
+    const rootStyle = html.style;
+    const bodyStyle = document.body.style;
     const scrollbarWidth = window.innerWidth - document.documentElement.clientWidth;
-    const scrollX = rootStyle.left ? parseFloat(rootStyle.left) : window.scrollX;
-    const scrollY = rootStyle.top ? parseFloat(rootStyle.top) : window.scrollY;
 
-    rootStyle.overflow = 'hidden';
+    let resizeRaf: number;
+    let scrollX: number;
+    let scrollY: number;
 
-    if (scrollbarWidth) {
-      rootStyle[paddingProp] = `${scrollbarWidth}px`;
-    }
-
-    // Only iOS doesn't respect `overflow: hidden` on document.body, and this
-    // technique has fewer side effects.
-    if (isIOS()) {
-      // iOS 12 does not support `visualViewport`.
+    function lockScroll() {
       const offsetLeft = window.visualViewport?.offsetLeft || 0;
       const offsetTop = window.visualViewport?.offsetTop || 0;
+      scrollX = rootStyle.left ? parseFloat(rootStyle.left) : window.scrollX;
+      scrollY = rootStyle.top ? parseFloat(rootStyle.top) : window.scrollY;
+
+      activeLocks.add(lockId!);
+
+      // We don't need to lock the scroll if there's already an active lock. However, it's possible
+      // that the one that originally locked it doesn't get cleaned up last. In that case, one of the
+      // newer locks needs to perform the style and scroll restoration.
+      if (activeLocks.size > 1) {
+        return cleanup;
+      }
+
+      originalStyles = {
+        position: rootStyle.position,
+        top: rootStyle.top,
+        left: rootStyle.left,
+        right: rootStyle.right,
+        ...(scrollbarWidth && {
+          overflowX: rootStyle.overflowX,
+          overflowY: rootStyle.overflowY,
+        }),
+      };
+      originalBodyOverflow = bodyStyle.overflow;
+
+      bodyStyle.overflow = 'hidden';
 
       Object.assign(rootStyle, {
-        position: 'fixed',
+        position: html.scrollHeight > html.clientHeight ? 'fixed' : '',
         top: `${-(scrollY - Math.floor(offsetTop))}px`,
         left: `${-(scrollX - Math.floor(offsetLeft))}px`,
         right: '0',
+        ...(scrollbarWidth && {
+          overflowY: html.scrollHeight > html.clientHeight ? 'scroll' : 'hidden',
+          overflowX: html.scrollWidth > html.clientWidth ? 'scroll' : 'hidden',
+        }),
```

Comment:
Any reason the styles are now contingent on `scrollbarWidth`?

The `scrollbarWidth` variable doesn't seem to recompute correctly when downsizing from viewport height that fits all contents to a viewport height that has a scrollbar

https://github.com/user-attachments/assets/3e168495-af56-4d96-b2e6-b9db18d83497

(Works fine the other way around)

Reply:
On iOS, it kind of scrolls the document html in the background in a strange way. It could be limited to use a platform check instead.

Reply:
Sounds good, should we leave a couple comments about these weird cases we are handling for the historical context? Like this one, and the one with conditional `position: fixed` that handles the scrollbar gutter bug in Chrome

Reply:
Added some comments and moved the variable inside the function so it should stay fresh